File: tests/performance/scenarios/mandelbrot/handler.js
module.exports = {
  main: async function (event, context) {
    const data = event.extensions.request.headers;
    width = data['width'];
    height = data['height'];
    iterations = data['iterations'];
    zoom = data['zoom'];

    console.log(`Received request with width=${width}, height=${height}, iterations=${iterations}, zoom=${zoom}`);

    return generateMandelbrotJson(width, height, iterations, zoom);
  }
}

function generateMandelbrotJson(width, height, iterations, zoom) {
  const safeWidth = Number.parseInt(width);
  const safeHeight = Number.parseInt(height);
  const maxIterations = Number.parseInt(iterations);
  const safeZoom = Number.parseFloat(zoom);
  const values = [];

  const xOffset = -0.7;
  const yOffset = 0;

  for (let y = 0; y < safeHeight; y++) {
    const row = [];
    const ci = (y / safeHeight - 0.5) / safeZoom + yOffset;

    for (let x = 0; x < safeWidth; x++) {
      const cr = (x / safeWidth - 0.5) / safeZoom + xOffset;
      row.push(mandelbrotEscapeIterations(cr, ci, maxIterations));
    }

    values.push(row);
  }

  return {
    type: "mandelbrot",
    width: safeWidth,
    height: safeHeight,
    maxIterations: maxIterations,
    zoom: safeZoom,
    values: values
  };
}

function mandelbrotEscapeIterations(cr, ci, maxIterations) {
  let zr = 0;
  let zi = 0;
  let iter = 0;

  while (iter < maxIterations && zr * zr + zi * zi < 4) {
    const nextZr = zr * zr - zi * zi + cr;
    zi = 2 * zr * zi + ci;
    zr = nextZr;
    iter++;
  }

  return iter;
}

File: tests/performance/scenarios/mandelbrot/handler.py
import json

def main(event, context):
    data = event['extensions']['request'].headers
    width = data.get("width")
    height = data.get("height")
    iterations = data.get("iterations")
    zoom = data.get("zoom")

    print(f"Received request with width={width}, height={height}, iterations={iterations}, zoom={zoom}")

    return generate_mandelbrot_json(width, height, iterations, zoom)


def generate_mandelbrot_json(width, height, iterations, zoom):
    safe_width = int(str(width).strip())
    safe_height = int(str(height).strip())
    max_iterations = int(str(iterations).strip())
    safe_zoom = float(str(zoom).strip())
    values = []

    x_offset = -0.7
    y_offset = 0.0

    for y in range(safe_height):
        row = []
        ci = (y / safe_height - 0.5) / safe_zoom + y_offset

        for x in range(safe_width):
            cr = (x / safe_width - 0.5) / safe_zoom + x_offset
            row.append(mandelbrot_escape_iterations(cr, ci, max_iterations))

        values.append(row)

    return {
        "type": "mandelbrot",
        "width": safe_width,
        "height": safe_height,
        "maxIterations": max_iterations,
        "zoom": safe_zoom,
        "values": values,
    }


def mandelbrot_escape_iterations(cr, ci, max_iterations):
    zr = 0.0
    zi = 0.0
    it = 0

    while it < max_iterations and (zr * zr + zi * zi) < 4:
        next_zr = zr * zr - zi * zi + cr
        zi = 2 * zr * zi + ci
        zr = next_zr
        it += 1

    return it
